[sw/device/lib/base] Prepare to add C warning flags

dmcardle opened the pull request.

dmcardle commented:
Ping @engdoreis :)

engdoreis reviewed (approved):
Sorry @dmcardle for the delay. 
Thanks for the PR, It LGTM.

I think @luismarques should have a look as well.

luismarques commented:
I haven't been following these warning fixing PRs very closely. One thing that I was concerned about from skimming other PRs in this series is that it wasn't super obvious from skimming them that the changes were correct. Taking the example of this specific PR, where you're casting the types to the signed counterpart, I can see it being easy to unintentionally change it to a different type altogether (if not now, perhaps later when things are refactored). Perhaps it's worth considering doing this in a more generic way, to avoid potential errors. Perhaps something like:

```c
#define SIGNED(value) _Generic((value), \
    uint8_t: (int8_t)(value), \
    uint16_t: (int16_t)(value), \
    uint32_t: (int32_t)(value), \
    uint64_t: (int64_t)(value), \
    _Static_assert(false, "Unsupported type for SIGNED") \
)

return ct_sltz32(SIGNED(~a & (a - 1)));
```

What do you think?

luismarques commented on sw/device/lib/base/macros.h:
```diff
@@ -441,6 +441,31 @@
+      uint16_t: (int16_t)(value), \
+      uint32_t: (int32_t)(value), \
+      uint64_t: (int64_t)(value), \
+      default: (struct OtSignConversionUnsupportedType){.err = 1})
```
> @luismarques I don't believe it's possible to use `static_assert` here because it will be evaluated regardless of which type matches `value`. Dynamic `CHECK` is undesirable because this is a compile-time error. So I introduced a struct (must have one member due to `-Wgnu-empty-struct`) that we can use when we want to emit a value that cannot be interpreted as an integer. WDYT?

Sounds good! That's the sort of thing I meant with the "some kind of" in `some_kind_of_assert` :)

luismarques reviewed (approved):
LGTM

dmcardle commented:
Oh, this is a little bit of a headache for two reasons:

1. Bindgen complained about parsing the `OtSignConversionUnsupportedType` struct. This is solved by wrapping the struct and sign conversion macros with `#ifndef RUST_PREPROCESSOR_EMIT`.
2. C++ unit tests pull in macros.h transitively, and the C++ compiler barfs when it hits the C generic. I may need to define a unittest-only C++ implementation for OT_SIGNED/OT_UNSIGNED.

dmcardle commented:
I've addressed the issues mentioned in the last comment with `#ifndef RUST_PREPROCESSOR_EMIT` and C++-specific implementations of `OT_SIGNED` and `OT_UNSIGNED`.

dmcardle commented:
Just a teensy bit more C++ magic. I added an `extern "C++"` block to macros.h.


It turns out that [print_unittest.cc](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/runtime/print_unittest.cc;l=7-9;drc=32355ad9c593d2b323a5c7e5ecaa975919ab9ae3) transitively includes macros.h inside an `extern "C"` block, but templates require C++ linkage.


https://dev.azure.com/lowrisc/opentitan/_build/results?buildId=117383&view=logs&j=5df5adf8-3577-5927-f0b2-31afed2697ee&t=1253ba27-4e8c-57af-d0ba-888237644875&l=4579

luismarques commented:
Sorry for all of the hassle @dmcardle, I hadn't imagined the yak shaving this would create :)
It's a little bit messier now but I would say it still LGTM. Hopefully there will be no more CI failures.

dmcardle commented:
Thanks, @luismarques! I think they are generally useful macros to have, so it's worth a bit of hassle. But I was surprised by the number of iterations, too — I don't think I've ever had to produce an `extern "C++"` block before!

dmcardle commented on sw/device/lib/base/macros.h:
```diff
@@ -443,4 +451,57 @@
  */
 #define OT_ALIGN_MEM(x) (uint32_t)(4 + (((uintptr_t)(x)-1) & ~3))
 
+#ifndef RUST_PREPROCESSOR_EMIT
```
Ah, both of these definitions probably need `#if !defined(__ASSEMBLER__) && !defined(NOSTDINC)` because test_rom_start.S includes macros.h.
